MTV-1352: Documentation of MTV performance 2.6 (#544)
kubev2v/forklift-documentation: documentation/modules/mtv-performance-recommendation.adoc

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: CONCEPT
[id="mtv-performance-recommendation_{context}"]
= {project-short} performance recommendations

> The purpose of this section is to show the current observed experience in the process of testing the performance and scale of {project-first}, to migrate VMs from vSphere to {virt}. 

Author: ```suggestion
The purpose of this section is to share recommendations for efficient and effective migration of virtual machines (VMs) using {project-first}, based on findings observed through testing.
```

* To perform fast migrations, VMware must have fast read access to datastores.  Networking between VMware ESXi hosts should be fast, ensure a 10 GiB network connection, and avoid network bottlenecks.

** Extend the VMware network to the OCP Workers Interface network environment.

** It is important to ensure that the VMware network offers high throughput (10 Gigabit Ethernet) and rapid networking to guarantee that the reception rates align with the read rate of the ESXi datastore.

** Be aware that the migration process uses significant network bandwidth and that the migration network is utilized. If other services utilize that network, it may have an impact on those services and their migration rates.

** For example, 200 to 325 MiB/s was the average network transfer rate from the `vmnic` for each ESXi host associated with transferring data to the OCP interface.

[id="mtv-performance-datastore-read_{context}"]
== Ensure fast datastore read speeds to ensure efficient and performant migrations.

Datastores read rates impact the total transfer times, so it is essential to ensure fast reads are possible from the ESXi datastore to the ESXi host.  

> Example in numbers: 200 to 300 MiB/s was the average read rate for both vSphere and ESXi endpoints for a single ESXi server. When more ESXi servers are used, high datastore read rates are possible.

Author: ```suggestion
Example in numbers: 200 to 300 MiB/s was the average read rate for both vSphere and ESXi endpoints for a single ESXi server. When multiple ESXi servers are used, higher datastore read rates are possible.
```

It is imperative to ensure that the network selected has connectivity to the OCP interface, has adequate bandwidth for migrations, and that the network interface is not saturated.

In environments with fast networks, such as 10GbE networks, migration network impacts can be expected to match the rate of ESXi datastore reads.

[id="mtv-performance-concurrent-disk_{context}"]
== Control maximum concurrent disk migrations per ESXi host.

Set the `MAX_VM_INFLIGHT MTV` variable to control the maximum number of concurrent VMs transfers allowed for the ESXi host. 

{project-short} allows for concurrency to be controlled using this variable; by default, it is set to 20.

When setting `MAX_VM_INFLIGHT`, consider the number of maximum concurrent VMs transfers are required for ESXi hosts. It is important to consider the type of migration to be transferred concurrently. Warm  migrations, which are defined by migrations of a running VM that will be migrated over a scheduled time.

> Warm migrations use snapshots to compare and migrate only the differences between previous snapshots of the disk.  The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to openshift occurs. 

Author: ```suggestion
Warm migrations use snapshots to compare and migrate only the differences between previous snapshots of the disk.  The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to {ocp-short} occurs. 
```

> In {project-short} 2.6, `MAX_VM_INFLIGHT` reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot.  The total set by `MAX_VM_INFLIGHT` is used to utilize one slot.

Author: ```suggestion
In {project-short} 2.6, `MAX_VM_INFLIGHT` reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot or the number of disks that belong to a single vm. The total set by `MAX_VM_INFLIGHT` is used to indicate how many concurrent VM tranfers per ESXi host is allowed.
```

.Example

* `MAX_VM_INFLIGHT = 20` and 2 ESXi hosts defined in the provider mean each host can transfer 20 VMs.

[id="mtv-performance-multiple-hosts_{context}"]
== Migrations complete faster using multiple hosts.

Using multiple hosts with registered VMs equally distributed among the ESXi hosts used for migrations leads to faster migration times.

Testing showed that when transferring more than 10 single disk VMS, each containing 35 GiB of data out of a total of 50G total, using an additional host can reduce migration time.

.Example

* 80 single disk VMs, containing 35 GiB of data each, using a single host took 2 hours and 43 minutes, with a migration rate of 294 MiB/s.

* 80 single disk VMs, containing 35 GiB of data each, using 8 ESXi hosts took 41 minutes, with a migration rate of 1,173 MiB/s.

[id="mtv-performance-bios-profile_{context}"]
== Set ESXi hosts BIOS profile and ESXi Host Power Management for High Performance

Where possible, ensure that hosts used to perform migrations are set with BIOS' profiles related to maximum performance.  Hosts which use Host Power Management controlled within vSphere should check that ‘High Performance’ is set.

> Testing showed that when transferring more than 10 VMs with both BIOS and host power management set accordingly, dataset read rates between tested VMs increased by 15 MiB.

Author: ```suggestion
Testing showed that when transferring more than 10 VMs with both BIOS and host power management set accordingly, migrations had an increase of 15 MiB in the average datastore read rate.
```

Comparing migrations of:

* 500 VMs using 8 ESXi hosts in 1 plan, `max_vm_inflight=100`,  took 5 hours and 10 minutes.

* 800 VMs using 8 ESXi hosts with 8 plans, `max_vm_inflight=100`, took 57 minutes.

Testing showed that by breaking one single large plan into multiple moderately sized plans, for example, 100 VMS per plan, the total migration time can be reduced.

[id="mtv-performance-max-values_{context}"]
== Maximum values tested

* Maximum number of ESXi hosts tested: 8
* Maximum number of VMs in a single migration plan: 500
* Maximum number of VMs migrated: 5000
* Maximum number of migration plans performed concurrently: 40
* Maximum single disk size migrated: 6 T disks, which contained 3 Tb of data
* Maximum number of disks on a single VM migrated: 50
* Highest observed single datastore read rate from a single ESXi server:  312 MiB/second
* Highest observed multi-datastore read rate using two ESXi servers and two datastores: 1,242 MiB/second
* Highest observed virtual NIC transfer rate to an {ocp-name} worker: 327 MiB/second

> * Maximum migration transfer rate of a single disk: 162 MiB/second (1.5 Tb case / time using utilized data)

Author: ```suggestion
* Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring 1.5Tb of non-zero data)
```

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: CONCEPT
[id="mtv-performance-recommendation_{context}"]
= {project-short} performance recommendations

The purpose of this section is to share recommendations for efficient and effective migration of virtual machines (VMs) using {project-first}, based on findings observed through testing.

> This document also aims to share key findings of what was observed through testing, and the summarized results of those tests, as shown by recommendations.

Reviewer: A similar disclaimer is needed for the abendum as well.

{project-short} allows for concurrency to be controlled using this variable; by default, it is set to 20.

When setting `MAX_VM_INFLIGHT`, consider the number of maximum concurrent VMs transfers are required for ESXi hosts. It is important to consider the type of migration to be transferred concurrently. Warm  migrations, which are defined by migrations of a running VM that will be migrated over a scheduled time.

Warm migrations use snapshots to compare and migrate only the differences between previous snapshots of the disk.  The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to {ocp-short} occurs. 

In {project-short} 2.6, `MAX_VM_INFLIGHT` reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot or the number of disks that belong to a single vm. The total set by `MAX_VM_INFLIGHT` is used to indicate how many concurrent VM tranfers per ESXi host is allowed.

.Example

* `MAX_VM_INFLIGHT = 20` and 2 ESXi hosts defined in the provider mean each host can transfer 20 VMs.

[id="mtv-performance-multiple-vms-concurrently_{context}"]

> == Migrations complete faster when migrating multiple VMs concurrently

Author: ```suggestion
== Migrations are completed faster when migrating multiple VMs concurrently
```

When multiple VMs from a specific ESXi host are to be migrated, starting concurrent migrations for multiple VMs leads to faster migration times. 

Testing demonstrated that migrating 10 VMs (each containing 35 GiB of data, with a total size of 50 GiB) from a single host is significantly faster than migrating the same number of VMs sequentially, one after another. 

It is possible to increase concurrent migration to more than 10 virtual machines from a single host, but it does not show a significant improvement. 

> .Example

Author: ```suggestion
.Examples
```

> * 1 single disk VMs took 6 minutes, with migration rate of 100 MiB/s

Author: ![image](https://github.com/user-attachments/assets/59fd6746-f456-4bec-a303-09aa0cb4bfd5)

Comparing migrations of:

* 500 VMs using 8 ESXi hosts in 1 plan, `max_vm_inflight=100`,  took 5 hours and 10 minutes.

* 800 VMs using 8 ESXi hosts with 8 plans, `max_vm_inflight=100`, took 57 minutes.

Testing showed that by breaking one single large plan into multiple moderately sized plans, for example, 100 VMS per plan, the total migration time can be reduced.

[id="mtv-performance-max-values_{context}"]
== Maximum values tested

* Maximum number of ESXi hosts tested: 8
* Maximum number of VMs in a single migration plan: 500
* Maximum number of VMs migrated in a single test: 5000
* Maximum number of migration plans performed concurrently: 40
* Maximum single disk size migrated: 6 T disks, which contained 3 Tb of data
* Maximum number of disks on a single VM migrated: 50
* Highest observed single datastore read rate from a single ESXi server:  312 MiB/second
* Highest observed multi-datastore read rate using eight ESXi servers and two datastores: 1,242 MiB/second
* Highest observed virtual NIC transfer rate to an {ocp-name} worker: 327 MiB/second

> * Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring 1.5Tb case/time using utilized data)

Reviewer 2: rate observed when transferring 1.5Tb case/time using utilized data - ether remove ether change to "nonconcurrent migration of  1.5Tb utilized data)

> * Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring 1.5Tb case/time using utilized data)

Author: ```suggestion
* Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring nonconcurrent migration of 1.5 Tb utilized data)
```

When setting `MAX_VM_INFLIGHT`, consider the number of maximum concurrent VMs transfers are required for ESXi hosts. It is important to consider the type of migration to be transferred concurrently. Warm  migrations, which are defined by migrations of a running VM that will be migrated over a scheduled time.

Warm migrations use snapshots to compare and migrate only the differences between previous snapshots of the disk.  The migration of the differences between snapshots happens over specific intervals before a final cut-over of the running VM to {ocp-short} occurs. 

In {project-short} 2.6, `MAX_VM_INFLIGHT` reserves one transfer slot per VM, regardless of current migration activity for a specific snapshot or the number of disks that belong to a single vm. The total set by `MAX_VM_INFLIGHT` is used to indicate how many concurrent VM tranfers per ESXi host is allowed.

.Example

* `MAX_VM_INFLIGHT = 20` and 2 ESXi hosts defined in the provider mean each host can transfer 20 VMs.

[id="mtv-performance-multiple-vms-concurrently_{context}"]
== Migrations complete faster when migrating multiple VMs concurrently

> Migrations complete faster when migrating multiple VMs concurrently. 

Author: ```suggestion
```

Comparing migrations of:

* 500 VMs using 8 ESXi hosts in 1 plan, `max_vm_inflight=100`,  took 5 hours and 10 minutes.

* 800 VMs using 8 ESXi hosts with 8 plans, `max_vm_inflight=100`, took 57 minutes.

Testing showed that by breaking one single large plan into multiple moderately sized plans, for example, 100 VMS per plan, the total migration time can be reduced.

[id="mtv-performance-max-values_{context}"]
== Maximum values tested

* Maximum number of ESXi hosts tested: 8
* Maximum number of VMs in a single migration plan: 500
* Maximum number of VMs migrated in a single test: 5000
* Maximum number of migration plans performed concurrently: 40
* Maximum single disk size migrated: 6 T disks, which contained 3 Tb of data
* Maximum number of disks on a single VM migrated: 50
* Highest observed single datastore read rate from a single ESXi server:  312 MiB/second
* Highest observed multi-datastore read rate using eight ESXi servers and two datastores: 1,242 MiB/second
* Highest observed virtual NIC transfer rate to an {ocp-name} worker: 327 MiB/second

> * Maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring nonconcurrent migration of 1.5 Tb utilized data)

Reviewer 2: Can you bold those below  3 as well 
-maximum migration transfer rate of a single disk: 162 MiB/second (rate observed when transferring nonconcurrent migration of 1.5 Tb utilized data)

 Maximum cold migration transfer rate of the multiple VMs (single disk) from a single ESXi host: 294 MiB/s (concurrent migration of 30 VMs, 35/50 GiB used, from Single ESXi)
* Maximum cold migration transfer rate of the multiple VMs (single disk) from multiple ESXi hosts: 1173MB/s (concurrent migration of 80 VMs, 35/50 GiB used, from 8 ESXi servers,
